src/mame/sega/suprloco.cpp

```diff
@@ -464,48 +594,71 @@ static const gfx_layout charlayout =
 
 
 static GFXDECODE_START( gfx_suprloco )
-	// sprites use colors 256-511 + 512-767
 	GFXDECODE_ENTRY( "chars", 0x6000, charlayout, 0, 16 )
 GFXDECODE_END
 
 
 void suprloco_state::suprloco(machine_config &config)
 {
 	// basic machine hardware
-	sega_315_5015_device &maincpu(SEGA_315_5015(config, m_maincpu, 4'000'000));   // 4 MHz (?)
-	maincpu.set_addrmap(AS_PROGRAM, &suprloco_state::main_map);
-	maincpu.set_addrmap(AS_OPCODES, &suprloco_state::decrypted_opcodes_map);
-	maincpu.set_vblank_int("screen", FUNC(suprloco_state::irq0_line_hold));
-	maincpu.set_decrypted_tag(":decrypted_opcodes");
-
-	Z80(config, m_audiocpu, 4'000'000);
+	SEGA_315_5015(config, m_maincpu, MASTER_CLOCK / 5); // see below
+	// this divider is actually /5 vs /6 depending on the /M1 line state from
+	//  the z80 during the 74LS161 counter reload; see system1.cpp which uses
+	//  an identical circuit.
+	m_maincpu->set_addrmap(AS_PROGRAM, &suprloco_state::main_map);
+	m_maincpu->set_addrmap(AS_OPCODES, &suprloco_state::decrypted_opcodes_map);
+	m_maincpu->set_vblank_int("screen", FUNC(suprloco_state::irq0_line_hold));
+	m_maincpu->set_decrypted_tag(":decrypted_opcodes");
+	m_maincpu->refresh_cb().set([this](offs_t offset, u8 data) {
+		m_m1_num = (m_m1_num + 1) % 5;
+		// if we've passed 5 M1 cycles, steal an additional
+		//  5x-20mhz-cycle-long cpu cycle. We do this at the third cycle to
+		//  average the correct speed.
+		if (m_m1_num == 2)
+			m_maincpu->adjust_icount(-1);
+	});
+
+	Z80(config, m_audiocpu, MASTER_CLOCK/5); // 4mhz
+	// This divide-by-five is done by a clever circuit with a 74LS390 in
+	//  bi-quinary mode.
 	m_audiocpu->set_addrmap(AS_PROGRAM, &suprloco_state::sound_map);
-	m_audiocpu->set_periodic_int(FUNC(suprloco_state::irq0_line_hold), attotime::from_hz(4 * 60));          // NMIs are caused by the main CPU
+	// Audio CPU INTs are caused by the falling edge of the V5 line counter
+	//  (4x per frame) and are auto-acked.
+	///TODO: once pals are dumped, make sure this uses the correct timing!
+	m_audiocpu->set_periodic_int(FUNC(suprloco_state::irq0_line_hold), attotime::from_hz(4 * 60));
```

Comment:
You can use a timer to trigger on specific scan lines.

Reply:
given that we don't yet have dumps of the pals which control the exact video timing, I don't know precisely which scanlines these fire on, especially since the flipscreen flag affects the pre-loaded value in the vertical and horizontal counters. The lines which these interrupts fire on are almost certainly different based on whether the screen is flipped or not. This requires further research.